doc: sysbuild documentation guide (#43846)
zephyrproject-rtos/zephyr: doc/develop/west/build-flash-debug.rst

  west build -t pristine

.. _west-multi-domain-builds:

> Multi-domain builds

Reviewer: I don't think this section placement makes sense, but GitHub won't let me upload the screenshot that shows what I mean. Please check Teams for the screenshot.

Author: ![image](https://user-images.githubusercontent.com/9802583/183830330-687bc2a4-f983-46bb-be14-c1b6edb34c64.png)

Author: Now looking like this:
![image](https://user-images.githubusercontent.com/9802583/184123325-b42fda1e-15fc-465a-beaf-169903f90508.png)

And the `west build` part for sysbuild has been split into a general description, and then a dedicated example.

Hope this is in line with your expectations.

The ``-o`` option is described further in the previous section.

.. _west-domains-file:

Domains file
------------

> A domains file specifies multiple domains which can be build, flashed, and

Reviewer 2: *built

Author: fixed.

debugged individually using the ``--domain <domain>`` argument.

The domains file is named ``domains.yaml`` and must be located in the top-level

> build folder of a multi-domain build.

Reviewer: The unclean separation between, and lack of consistency between, the sysbuild and west build documentation could use a bit of improvement, I think.

Here are some example issues in my view:

1. We're using "multi-domain build" here in this document, but that is not a defined term in the sysbuild documentation guide, so that's confusing.
2. This is also being placed under the "examples" section; but it doesn't belong there. Did you build the docs HTML and check the section placement?
3. A definition of the contents of the domains file really don't belong in the west build tool docs, IMO -- the file is generated by sysbuild, so the docs should appear in the sysbuild documentation

Please move the parts that belong to the sysbuild documentation to the sysbuild guide, and add a separate standalone section under the west build section here that documents the sysbuild integration, referring to the sysbuild docs as necessary. The software layering is cmake -> sysbuild -> west build, so the documentation layering should match that.

Author: > 1. We're using "multi-domain build" here in this document, but that is not a defined term in the sysbuild documentation guide, so that's confusing.

Good point.
Let me add that to the definition overview.
But I don't think inside sysbuild itself there is much need for using this term, cause it's not until you really want to point out a specific domain of the multiple we really need to differentiate.

> 2. This is also being placed under the "examples" section, but it doesn't belong there. Did you build the docs HTML and check the section placement?

I did build it and I placed it where it made sense compared to description / examples of other build options, but I didn't pay attention to how it looked in TOC, so I didn't notice the overall examples part.
Will move it.

> 3. A definition of the contents of the domains file really don't belong in the west build tool docs, IMO -- the file is generated by sysbuild, so the docs should appear in the sysbuild documentation

But the file is only used by `west`.
If you don't use `west` and only uses CMake + ninja, then you don't care for `domains.yml`.
I see this file as similar to `runners.yml` which I consider a `west` owned file, at least the format of the file is owned by `west`, but of course generated by the Zephyr build system here:
https://github.com/zephyrproject-rtos/zephyr/blob/236484456f30576271f91b157840f049afade918/cmake/flash/CMakeLists.txt#L111-L113
`runners.yaml` doesn't seem to be documented, but if it were to be documented I would argue that it belongs with west commands like flash and debug.

You may say we don't want to document domains.yaml, because runners.yaml is not documented, and I'm fine with that and can remove this part, but if we are to describe domains.yaml file then I think it belongs as part of above mentioned west commands.

Reviewer: > If you don't use west and only uses CMake + ninja, then you don't care for domains.yml.

OK, I didn't know that, and that's a fair point.

I'd say just remove the documentation for it for now so we can break it as often as we want in the early days. If it settles down and we want to make the format public, we can do it later. OK with you?

Author: > I'd say just remove the documentation for it for now so we can break it as often as we want in the early days.

Good plan :+1:

Author: > > I'd say just remove the documentation for it for now so we can break it as often as we want in the early days.
> 
> Good plan :+1:

Done

Author: > * We're using "multi-domain build" here in this document, but that is not a defined term in the sysbuild documentation guide, so that's confusing.
> 

Done.

> * This is also being placed under the "examples" section, but it doesn't belong there. Did you build the docs HTML and check the section placement?

Split into description and an example.
Description moved up and generally cleaned up.

The ``domains.yaml`` contains domains along with their build directory.
A ``default`` entry exists in the file.

> Commands which only handles a single domain, such as ``west debug``, uses this

Author: fixed.

entry to select the domain to use if no domain is specified.

> A sample ``domains.yaml`` file:

Reviewer 3: ```suggestion
The following is a sample ``domains.yaml`` file:
```

.. code-block:: yaml

  default: <domain-n>
  build_dir: <toplevel-build-dir>
  domains:
    - name: <domain-name-1>:
      build_dir: <build-dir-domain-1>
    - name: <domain-name-2>:
      build_dir: <build-dir-domain-2>

For example, the :ref:`hello_world` sample built with `MCUboot`_

> support, will look like:

Reviewer 3: ```suggestion
support will look as follows:
```

.. code-block:: yaml

  default: hello_world
  build_dir: <toplevel-build-dir>
  domains:
    - name: mcuboot
      build_dir: <toplevel-build-dir>/mcuboot
    - name: hello_world
      build_dir: <toplevel-build-dir>/hello_world

> Sysbuild

Reviewer: Same issue here; this is under "Examples" too. An additional issue is that the content in this section is not just about west build, but is about the other west extensions -- like `flash` and `debug` -- documented on this page. The structure here really doesn't make sense.

Author: Moved and cleaned up.

  west flash -H -r jlink

.. _west-multi-domain-flashing:

> multi-domain flashing

Reviewer: Please fix case and resolve inconsistent terminology

Author: tried to fix inconsistency, please check if you still find inconsistency in this section.

=====================

When a :ref:`west-multi-domain-builds` folder is detected, then ``west flash``
will flash all domains in the order they appear in the  :ref:`west-domains-file`.

It is possible to build a single domain in a multi-domain project by using
``--domain``.

For example, in a multi-domain build with :ref:`hello_world` and
`MCUboot`_, you can use the ``--domain hello_world`` domain to only flash
this domain::

>   west flash --domain hello_world

Reviewer: Some issues related to argument passing here.

sysbuild/index.rst says:

> When invoking ``west flash`` on a build consisting of multiple images, each
> image is flashed in sequence. Extra arguments such as ``--runner jlink`` are
> passed to each invocation.
>
> For more details, see :ref:`west-multi-domain-flashing`.

But the examples here are all about the `--domain` argument, not the `--runner` argument. I think that's confusing; the examples should be consistent so the reader knows they've found the details here.

Second problem: we discussed that some options, should not be handled agnostically by sysbuild-aware flash, such as things like --dev-id which are not implemented by all runners. I don't see documentation here reflecting the decisions you made about how those are handled.

Author: > But the examples here are all about the --domain argument, not the --runner argument.

I agree. Leftover from where I had most of the `west build|flash|debug` documentation inside sysbuild, before moving it to `west` page and decided to only have a brief mentioning of this together with a link to `west` docs.
Will clean up.

> Second problem: we discussed that some options, like --dev-id, should not be handled agnostically by sysbuild-aware flash.

I remember we briefly touched upon this but only decided that we be agnostic to all args, and just pass them on.
If a user wants to have, let's say `--dev-id` used for only a given domain, he can do so by also specifying such domain, like this: `west flash --dev-id <id> --domain <dom>`

Reviewer: > I remember we briefly touched upon this but only decided that we be agnostic to all args, and just pass them on.

That's odd, I remember the opposite. I thought you were going to have some sort of allowlist of arguments that you would allow that were guaranteed to be domain independent. Then you would fail the command if something outside the allowlist, and therefore potentially domain-specific, was given.

Author: I made a PR here: https://github.com/zephyrproject-rtos/zephyr/pull/48914

This will print a warning to the user if he does:
```
$ west flash --<runner-arg>
```
like
```
$ west flash --dev-id 12345
```
on a multi domain build.

But if user does:
```
$ west flash --dev-id 12345 --domain <domain>
```
no warning is printed.

The allows users to still apply runners args and let the user decide if it makes sense, while at the same time he is warned that that currently this feature is considered experimental.

Allows us to get some feedback on this feature.
Should we open a common issue that we could link to, so we collect experiences at a single place, like:
```
> WARNING: Specifying runner options for multiple domains is experimental.
> If problems are experienced, please specify a single domain using '--domain <domain>'         
> Any feedback on this feature can be given https://github.com/zephyrproject-rtos/zephyr/issues/<a_number>
```

  west debug -H -r jlink

.. _west-multi-domain-debugging:

Multi-domain debugging
======================

> ``west debug`` can only debug a single domain at a time, so when a
> :ref:`west-multi-domain-builds` folder is detected, then ``west debug``
> will debug the ``default`` domain specified in the :ref:`west-domains-file`.

Reviewer 3: ```suggestion
``west debug`` can only debug a single domain at a time. When a
:ref:`west-multi-domain-builds` folder is detected, ``west debug``
will debug the ``default`` domain specified in the :ref:`west-domains-file`.
```

> The default domain will be the application given by the source directory as::

Reviewer 3: ```suggestion
The default domain will be the application given as the source directory.
See the following example::
```

  west build --sysbuild path/to/source/directory

> For example, when building ``hello_world`` with `MCUboot`_ using sysbuild then
> ``hello_world`` becomes the default domain::

Reviewer 3: ```suggestion
For example, when building ``hello_world`` with `MCUboot`_ using sysbuild,
``hello_world`` becomes the default domain::
```

  west build --sysbuild samples/hello_world

So to debug ``hello_world`` you can do::

  west debug

or::

  west debug --domain hello_world

> if you wish to debug MCUboot, you must explicitly specify MCUboot as the domain

Reviewer 3: ```suggestion
If you wish to debug MCUboot, you must explicitly specify MCUboot as the domain
```

  west debug -H -r jlink

.. _west-multi-domain-debugging:

> Multi-domain debugging

Reviewer: Do we really want to be recommending use of `west debug` here? I think that in a multi-image context, `west attach` is probably the better command to be using. Since `attach` doesn't reflash anything, it can't mess anything up, while `debug` can reflash things, which might result in odd states on flash if domains need to be flashed in a specific sequence.

Author: but if you're working on a sample that enables MCUBoot, then you probably want to have the application re-flashed before debugging if you made code changes.

Which is also the reason the main application (which is the one you pointed to at build time) is the application that default will be used for debugging.

I don't claim to know the right answer here, cause it's hard to predict exactly how sysbuild is used for those purposes, but this behavior is somewhat similar to the existing behavior in the nRF Connect SDK multi image.
And I haven't heard complaints on that part. So instead of recommending two different approaches, one approach when not using sysbuild, another when using sysbuild, then I felt it safest to just keep the existing way.

Reviewer: > And I haven't heard complaints on that part.

I have had to help at least one user who got confused about this. But it's OK, let's leave it as is for now and we can adjust depending on the complaints we get from users.

Author: I've been thinking a bit.

I still think it makes sense to describe how `west debug` functions, cause if that is what users are used to, it's probably a workflow they will continue to use.

But I think we can address the surprises of `west debug` in a multi domain setup, and instead highlight the possibility of `west attach` for this scenario as a good (better !?) alternative in this scenario.

But I will focus on the other comments first, and the `attach` addition can be introduced later if need be.
